aten/src/ATen/native/TopKImpl.h

```diff
@@ -49,36 +49,60 @@ void topk_impl_loop(
       if (largest) {
         std::partial_sort(queue.begin(), queue.begin() + k, queue.end(),
           [](const elem_t& x, const elem_t& y) -> bool {
-            return ((_isnan<accscalar_t>(x.first) && !_isnan<accscalar_t>(y.first)) || (x.first > y.first));
+            if ((_isnan<accscalar_t>(x.first) && !_isnan<accscalar_t>(y.first)) || (x.first > y.first)) return true;
+            if ((!_isnan<accscalar_t>(x.first) && _isnan<accscalar_t>(y.first)) || (x.first < y.first)) return false;
+            return x.second < y.second;
           });
       } else {
         std::partial_sort(queue.begin(), queue.begin() + k, queue.end(),
           [](const elem_t& x, const elem_t& y) -> bool {
-            return ((!_isnan<accscalar_t>(x.first) && _isnan<accscalar_t>(y.first)) || (x.first < y.first));
+            if ((!_isnan<accscalar_t>(x.first) && _isnan<accscalar_t>(y.first)) || (x.first < y.first)) return true;
+            if ((_isnan<accscalar_t>(x.first) && !_isnan<accscalar_t>(y.first)) || (x.first > y.first)) return false;
+            return x.second < y.second;
           });
       }
     } else {
       if (largest) {
-        std::nth_element(queue.begin(), queue.begin() + k - 1, queue.end(),
+        if (stable) {
+          std::sort(queue.begin(), queue.end(),
```

Comment:
Should you call `std::stable_sort` instead?

Reply:
`std::stable_sort` works as well but with higher complexity. O(N·log(N)^2) --vs-- O(N·log(N)) for  `std::stable_sort` and `std::sort` respectively. Since we already keep the idx of each element as the second value of `elem_t`, we can stable sort with `std::sort` at lower cost. Looks like tensorflow also mentions this: https://github.com/tensorflow/tensorflow/blob/v2.10.0/tensorflow/core/kernels/topk_op.cc#L171-L172.

Reply:
Oh, make sense, you checked both values and indices now.

Reply:
> Hmm, why can't we directly dispatch to at::sort which already handles stable=True? Then we get CUDA support for free as well.

@nikitaved, I see the `at::sort` kernel uses `std::stable_sort` for stable sort. Due to the above comment on complexities between `std::stable_sort` --vs-- `std::sort`, do you think it's better to use `std::sort` since we already keep the idx of each element for topk?

Reply:
We can at least dispatch into `at::sort` for CUDA, I think. Also, we can potentially replace `stable_sort` with `sort` handling composite keys, so, in a way, I would assume that dispatching into `at::sort` is better, and then we could improve `at::sort` as a follow-up.

Reply:
Hi @nikitaved~ let me know if any update on this PR, thanks!

Reply:
Hey! Sorry for the delay, now I am back! Let's first just implement the stable flag for both cpu and cuda that dispatches to `sort`. Then we can improve `at::sort` for the CPU via just replaceing `std::stable_sort` with `std::sort` and an updated comparator? What do you think?

Reply:
Hi @nikitaved ~ Sounds good to me! I will update this PR as soon as ready, thanks.